### PR TITLE
Implements class based views for transactions using API View

### PR DESCRIPTION
**Pull request by Hamza-Chaudhary65**
- Implements class based views for transactions using API View
- Uses API URL Versioning

### CODE REVIEW

File: fintrack/transactions/v2/views.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+from transactions.models import Transaction
+from transactions.serializer import TransactionSerializer
+from transactions.transaction_helper import TransactionManager
+
+transaction_manager = TransactionManager()
+
+
+class TransactionView(APIView):
+
+    permission_classes = [IsAuthenticated]
```
Comment (sheralim012) [inline fintrack/transactions/v2/views.py]: update drf settings instead

File: fintrack/accounts/migrations/0005_alter_account_holder.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Generated by Django 4.2.13 on 2024-07-18 19:20
```
Comment (sheralim012) [inline fintrack/accounts/migrations/0005_alter_account_holder.py]: always minimize the number of migrations in a PR

File: fintrack/transactions/v2/views.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        serializer = TransactionSerializer(transaction)
+        return Response(serializer.data)
+
+
+class RevertTransactionView(APIView):
+
+    permission_classes = [IsAuthenticated]
+
+    def post(self, request, pk):
+        response = None
```
Comment (sheralim012) [inline fintrack/transactions/v2/views.py]: is this necessary?

Comment (sheralim012) [inline fintrack/transactions/v2/views.py]: is this necessary?

File: fintrack/accounts/migrations/0005_alter_account_holder.py (hunk 2 above)
Comment (sheralim012) [inline fintrack/accounts/migrations/0005_alter_account_holder.py]: always minimize the number of migrations in a PR

File: fintrack/transactions/v2/views.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+        transaction = get_object_or_404(Transaction, pk=pk, transaction_from_account__holder=request.user)
+
+        if transaction.amount < 0:
```
Comment (sheralim012) [inline fintrack/transactions/v2/views.py]: it this to check if transaction is already a revert?

File: fintrack/transactions/v2/views.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            response = Response(
+                "Can not revert this transaction", 
+                status=status.HTTP_400_BAD_REQUEST
+            )
+        else:
+            transaction_manager.handle_revert_transaction(transaction, request.data)
+            response = Response("Reverted Successfully")
```
Comment (sheralim012) [inline fintrack/transactions/v2/views.py]: api's return json response, and valid json has key-value pairs

Comment (sheralim012) [inline fintrack/transactions/v2/views.py]: api's return json response, and valid json has key-value pairs

File: fintrack/transactions/v2/views.py (hunk 4 above)
Comment (sheralim012) [inline fintrack/transactions/v2/views.py]: it this to check if transaction is already a revert?

File: fintrack/transactions/v2/views.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        paginator = PageNumberPagination()
+        paginated_transactions = paginator.paginate_queryset(transactions, request)
+
+        serializer = TransactionSerializer(paginated_transactions, many=True)
+        return paginator.get_paginated_response(serializer.data)
+
+    def post(self, request):
+        response = None
+
+        if request.data["amount"] < 0:
```
Comment (sheralim012) [inline fintrack/transactions/v2/views.py]: it's it a validation and should be in serializer?

File: fintrack/transactions/v2/views.py (hunk 3 above)
Comment (sheralim012) [inline fintrack/transactions/v2/views.py]: is this necessary?

File: fintrack/transactions/v2/views.py (hunk 4 above)
Comment (sheralim012) [inline fintrack/transactions/v2/views.py]: it this to check if transaction is already a revert?

File: fintrack/transactions/v2/views.py (hunk 5 above)
Comment (sheralim012) [inline fintrack/transactions/v2/views.py]: api's return json response, and valid json has key-value pairs

File: fintrack/transactions/v2/views.py (hunk 6 above)
Comment (sheralim012) [inline fintrack/transactions/v2/views.py]: it's it a validation and should be in serializer?

File: fintrack/accounts/migrations/0005_alter_account_holder.py (hunk 2 above)
Comment (sheralim012) [inline fintrack/accounts/migrations/0005_alter_account_holder.py]: always minimize the number of migrations in a PR

**Review by sheralim012 — CHANGES_REQUESTED** [review]

File: fintrack/transactions/v2/views.py (hunk 4 above)
Comment (Hamza-Chaudhary65) [reply fintrack/transactions/v2/views.py]: Yes

File: fintrack/accounts/migrations/0001_initial.py [hunk 7]
```diff
@@ -1,4 +1,4 @@
-# Generated by Django 4.2.13 on 2024-07-08 18:50
+# Generated by Django 4.2.13 on 2024-07-22 09:13
```
Comment (sheralim012) [inline fintrack/accounts/migrations/0001_initial.py]: we shouldn't be changing previous migrations

File: fintrack/transactions/v2/views.py (hunk 4 above)
Comment (sheralim012) [reply fintrack/transactions/v2/views.py]: wouldn't it be better to have it as type/category or something like that to denote that it's a revert transaction

**Review by sheralim012 — CHANGES_REQUESTED** [review]

File: .vscode/settings.json
Comment (sheralim012) [inline .vscode/settings.json:1]: ignore this directory

File: fintrack/transactions/serializer.py [hunk 8]
```diff
@@ -7,3 +7,10 @@ class TransactionSerializer(serializers.ModelSerializer):
     class Meta:
         model = Transaction
         fields = '__all__'
+
+    def validate_amount(self, amount):
+        if amount < 0:
+            raise serializers.ValidationError("Transaction Can not be negative")
+    
+    
+
```
Comment (sheralim012) [inline fintrack/transactions/serializer.py]: unnecessary extra empty lines

File: fintrack/transactions/transaction_helper.py [hunk 9]
```diff
@@ -54,3 +55,8 @@ def update_balance_on_transaction_creation(self, account, transaction_updated_da
                 account, 
                 transaction_updated_data["amount"]
             )
+        
+    def get_json_response(self, message):
```
Comment (sheralim012) [inline fintrack/transactions/transaction_helper.py]: I told you to avoid making one-liner function

**Review by sheralim012 — CHANGES_REQUESTED** [review]

File: fintrack/transactions/transaction_helper.py (hunk 9 above)
Comment (Hamza-Chaudhary65) [reply fintrack/transactions/transaction_helper.py]: Created it just to improve readability.

**Review by sheralim012 — APPROVED** [review]

**Pull request merged**